packages/test-kit/src/sync-utils-contract.ts

```diff
@@ -0,0 +1,63 @@
+import { expect } from 'chai';
+import type { IFileSystemSync } from '@file-services/types';
+import type { ITestInput } from './types';
+
+export function syncUtilsContract(testProvider: () => Promise<ITestInput<IFileSystemSync>>): void {
+  describe('SYNC utilities contract', () => {
+    let testInput: ITestInput<IFileSystemSync>;
+
+    beforeEach(async () => (testInput = await testProvider()));
+    afterEach(async () => await testInput.dispose());
+
+    describe('ensureDirectorySync', () => {
+      it(`creates intermediate directories`, () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.populateDirectorySync(tempDirectoryPath, { a: {} });
+        const newDirPath = fs.join(tempDirectoryPath, 'animals', 'mammals', 'chiroptera');
+        fs.ensureDirectorySync(newDirPath);
+
+        expect(fs.directoryExistsSync(newDirPath)).to.equal(true);
+      });
+
+      it(`succeeds if directory already exists, preserves its contents`, () => {
+        const { fs, tempDirectoryPath } = testInput;
+        fs.populateDirectorySync(tempDirectoryPath, { animals: { mammals: { 'vesper-bat.txt': '🦇' } } });
+        fs.ensureDirectorySync(fs.join(tempDirectoryPath, 'animals', 'mammals'));
+
+        const filePath = fs.join(tempDirectoryPath, 'animals', 'mammals', 'vesper-bat.txt');
+        expect(fs.readFileSync(filePath, 'utf8')).to.equal('🦇');
+      });
+
+      it(`throws when attempting to overwrite existing file`, () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.populateDirectorySync(tempDirectoryPath, { 'vesper-bat.txt': '🦇' });
+        const newDirPath = fs.join(tempDirectoryPath, 'vesper-bat.txt');
+
+        expect(() => fs.ensureDirectorySync(newDirPath)).to.throw('EEXIST');
+      });
+
+      it(`throws when attempting to create a directory inside of a file`, () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.populateDirectorySync(tempDirectoryPath, { 'vesper-bat.txt': '🦇' });
+        const newDirPath = fs.join(tempDirectoryPath, 'vesper-bat.txt', 'habitat');
+
+        expect(() => fs.ensureDirectorySync(newDirPath)).to.throw('ENOTDIR');
+      });
+
+      it('handles special paths gracefully', () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.chdir(tempDirectoryPath);
+        expect(() => fs.ensureDirectorySync('.')).to.not.throw();
+        expect(() => fs.ensureDirectorySync('..')).to.not.throw();
+
+        // This doesn't match node fs, which considers '' to be different from '.',
```

Comment:
memory fs should follow suite rather than being tested separately.
it's a matter of changing `resolvePath` (return '' back as '') and getNode (return null for '')

Reply:
I'm pretty sure it shouldn't be fixed at the level of `resolvePath`:

<img src="https://user-images.githubusercontent.com/415252/88364686-5d245880-cd8c-11ea-8843-99f74fb62435.png" width="407">

Reply:
I removed this `ensureDirectorySync('')` assertion for now, as I'm not yet sure how many changes throughout it will require, and don't want to hold up this PR.

Just to be clear, `ensureDirectorySync('')` doesn't match Node's `fs.mkdirSync('')` behavior in both memory fs and node fs.